core/domain/classifier_services.py

```diff
@@ -127,6 +127,52 @@ def classify_string_classifier_rule(state, normalized_answer):
     return None
 
 
+def create_classifier_training_jobs(exploration, state_names):
+    """Creates ClassifierTrainingJobModel instances for all the state names
+    passed into the function.
+
+    Args:
+        exploration: Exploration. The Exploration Domain object.
+        state_names: list. List of state names.
+    """
+    exp_id = exploration.id
+    exp_version = exploration.version
+    states = exploration.states
+    for state_name in state_names:
+        state = states[state_name]
+        training_data = state.get_training_data()
+        interaction_id = state.interaction.id
+        algorithm_id = feconf.INTERACTION_CLASSIFIER_MAPPING[
+            interaction_id]['algorithm_id']
+        create_classifier_training_job(algorithm_id, interaction_id,
+                                       exp_id, exp_version, state_name,
+                                       training_data,
+                                       feconf.TRAINING_JOB_STATUS_NEW)
+
+
+def update_classifier_exploration_mappings(exploration, state_names):
+    """Creates new ClassifierExplorationMappingModel instances for all the
+    state names passed into the function.
+
+    Args:
+        exploration: Exploration. The Exploration Domain object.
+        state_names: list. List of state names.
+    """
+    exp_id = exploration.id
```

Comment:
OK, I think we need to restructure this logic so that it is _not_ confusing. Can you write pseudocode for it in a way that's as simple and clear to the reader as possible? Let's try and agree on that, then you can implement it fully. This is something you might want to work through with @anmolshkl and @prasanna08 as well, in order to make absolutely sure you are all on the same page.

There are a couple of other comments I have:
- When an exploration is reverted to an earlier version, we should just reuse the classifiers from that earlier version (or create new ones if they don't exist).
- The above point illustrates why it's kind of important to have classifiers for each version, if possible, so I'm not sure why we would delete v1's job. Is there a way to detect that v1's job is in the queue and still map v2 to whatever the output of v1's job is when it comes back? Otherwise you're just deleting a job and replacing it with something that's further back in the queue.